numpy/core/src/npymath/ieee754.cpp

```diff
@@ -135,129 +146,140 @@ static float _nextf(float x, int p)
 }
 
 #if defined(HAVE_LDOUBLE_DOUBLE_DOUBLE_BE) || \
-    defined(HAVE_LDOUBLE_DOUBLE_DOUBLE_LE)
+        defined(HAVE_LDOUBLE_DOUBLE_DOUBLE_LE)
 
 /*
  * FIXME: this is ugly and untested. The asm part only works with gcc, and we
  * should consolidate the GET_LDOUBLE* / SET_LDOUBLE macros
  */
-#define math_opt_barrier(x) \
-        ({ __typeof (x) __x = x; __asm ("" : "+m" (__x)); __x; })
-#define math_force_eval(x) __asm __volatile ("" : : "m" (x))
+#define math_opt_barrier(x)    \
+    ({                         \
+        __typeof(x) __x = x;   \
+        __asm("" : "+m"(__x)); \
+        __x;                   \
+    })
+#define math_force_eval(x) __asm __volatile("" : : "m"(x))
 
 /* only works for big endian */
-typedef union
-{
+typedef union {
     npy_longdouble value;
-    struct
-    {
+    struct {
         npy_uint64 msw;
         npy_uint64 lsw;
     } parts64;
-    struct
-    {
+    struct {
         npy_uint32 w0, w1, w2, w3;
     } parts32;
 } ieee854_long_double_shape_type;
 
 /* Get two 64 bit ints from a long double.  */
 
-#define GET_LDOUBLE_WORDS64(ix0,ix1,d) \
-do {                                   \
-  ieee854_long_double_shape_type qw_u; \
-  qw_u.value = (d);                    \
-  (ix0) = qw_u.parts64.msw;            \
-  (ix1) = qw_u.parts64.lsw;            \
-} while (0)
+#define GET_LDOUBLE_WORDS64(ix0, ix1, d)     \
+    do {                                     \
+        ieee854_long_double_shape_type qw_u; \
+        qw_u.value = (d);                    \
+        (ix0) = qw_u.parts64.msw;            \
+        (ix1) = qw_u.parts64.lsw;            \
+    } while (0)
 
 /* Set a long double from two 64 bit ints.  */
 
-#define SET_LDOUBLE_WORDS64(d,ix0,ix1) \
-do {                                   \
-  ieee854_long_double_shape_type qw_u; \
-  qw_u.parts64.msw = (ix0);            \
-  qw_u.parts64.lsw = (ix1);            \
-  (d) = qw_u.value;                    \
-} while (0)
+#define SET_LDOUBLE_WORDS64(d, ix0, ix1)     \
+    do {                                     \
+        ieee854_long_double_shape_type qw_u; \
+        qw_u.parts64.msw = (ix0);            \
+        qw_u.parts64.lsw = (ix1);            \
+        (d) = qw_u.value;                    \
+    } while (0)
 
-static npy_longdouble _nextl(npy_longdouble x, int p)
+static long double
+_next(long double x, int p)
 {
-    npy_int64 hx,ihx,ilx;
+    npy_int64 hx, ihx, ilx;
     npy_uint64 lx;
     npy_longdouble u;
+    constexpr npy_longdouble eps = 2.465190328815662e-32; // 0x1.0000000000000p-105L
```

Comment:
I wonder if a test for his makes sense?  But seems fine to skip.  I don't think this is normally used.

Reply:
I checked and the internal representation of the two values is the same as I commited them.

Reply:
But to avoid confusion, I updated the patch with the 'L'-suffixed value :-)

Reply:
How did you validate it?
```
In [1]: np.longdouble("2.465190328815662e-32") - 2.465190328815662e-32
Out[1]: 1.07979694287724704104e-48
```
And this is 80bit extended precision longdouble.  IBM double-double has a much bigger mantissa.  Probably this function isn't even needed, but I am not 100% sure (IIRC aarch64 or so may not provide `nextafter` even though it is C99).

Maybe use `exp2l` to calculate it?  That is how we do it for `np.finfo` as well (in Python).

Reply:
https://godbolt.org/z/EbaqdfPsP

Reply:
I am worried that this is aarch64 (or similar): https://godbolt.org/z/53f8j4jY7

Reply:
ohhh, I got it the wrong way.  It actually must be _without_ the `L`, because than it is the correct double precision value (i.e. exact in binary).  But with the `L` it is not an exact in binary.

Reply:
indeed. Let's do a math call then.